### ISSUE TITLE
RejectedExecutionException

### ISSUE THREAD

**Issue by gsalgun** [pusher/pusher-websocket-java#107]
This started appearing on crash logs after updating to the last version. 

```

Fatal Exception: java.util.concurrent.RejectedExecutionException: Task com.pusher.client.util.Factory$1@345609e9 rejected from java.util.concurrent.ThreadPoolExecutor@12dd666e[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
       at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2011)
       at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:793)
       at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1334)
       at java.util.concurrent.Executors$DelegatedExecutorService.execute(Executors.java:583)
       at com.pusher.client.util.Factory.queueOnEventThread(Factory.java:101)
       at com.pusher.client.connection.websocket.WebSocketConnection.connect(WebSocketConnection.java:65)
       at com.pusher.client.Pusher.connect(Pusher.java:175)
       at com.forplay.app.PusherService.startPusher(PusherService.java:171)
       at com.forplay.app.PusherService.attemptToReconnect(PusherService.java:221)
       at com.forplay.app.PusherService.access$500(PusherService.java:46)
       at com.forplay.app.PusherService$2.run(PusherService.java:203)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:5832)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

**Comment by gsalgun** [comment]
Hello, 
Is there any update on this? This is the most frequent crash in our crash reports. Previously we had
https://github.com/pusher/pusher-websocket-java/issues/54

Latest update solved that one but introduced this current crash. Please give us a feedback on its state, or help us work around this.

**Comment by jpatel531** [comment]
@mdpye - it looks like the old race condition in a different form. it's likely that it's occurring 'between' these two lines ( https://github.com/pusher/pusher-websocket-java/blob/master/src/main/java/com/pusher/client/util/Factory.java#L113-L114 ), which is plausible if the `shutdown()` method is taking a while to finish. so during the race condition, one thread is getting the shutting down eventqueue, while the other is shutting it down and about to nullify it. 

any thoughts? it's not neat but one way i can think of is not to nullify eventQueue in `shutdownThreads()`, but replace any shutdown ones with a new one in `getEventQueue()`

**Comment by mdpye** [comment]
`getEventQueue` and `shutdownThreads` are both `synchronized` methods, which is equivalent to them being wrapped in `synchronized (this)`, so there can't be a race there.

However, `queueOnEventThread` is not synchronized, so while it is guaranteed to get a consistent result when it calls `getEventQueue`, someone else could call `shutdownThreads` (and therefore shut down the event queue) after it has been returned to `queueOnEventThread`, but before `execute` is called on it.

**Comment by gsalgun** [comment]
Hello @mdpye, @jpatel531 ,

Is there any further development on this?

**Comment by jpatel531** [comment]
Hi @gsalgun - sorry for the lack of response. I've now got time to tackle this, and am working on a fix

**Comment by jpatel531** [comment]
Hey @gsalgun - I've just published 1.1.3 - hopefully this should resolve the issue. Thank you for your patience

**Comment by gsalgun** [comment]
Hi @jpatel531 
Thank you for the fix, I will update ASAP and let you know if it has disappeared from logs
